### ISSUE TITLE
Init-db container doesn't see posgresql

### ISSUE THREAD

**Issue by TheFuzzyTech** [mvitale1989/helm-taiga#7]
I'm relatively new to kubernetes so this is more than likely a 'me' problem. I'm struggling to get this chart running. When the init-db container runs it never see's the db as up. I've tried setting dbHost to ${release_name}-postgresql and I've also tried just entering the value of the pod name manually (release-postresql-0). I'm not sure what else could be wrong at this point. I'm trying to run this in a local k8s cluster. Below is a sanitized values.yaml

```yaml
### Default values for taiga.

image:
  repository: mvitale1989/docker-taiga
  tag: "20191031-4.2.14"
  initRepository: alpine
  initTag: 3.7
  pullPolicy: IfNotPresent

service:
  type: ClusterIP
  port: 80

ingress:
  enabled: true
  annotations: 
    kubernetes.io/tls-acme: "true"
    # kubernetes.io/ingress.class: nginx
    # kubernetes.io/tls-acme: "true"
  hosts:
  - taiga.org.com
  tls: 
  - secretName: taiga-tls
    hosts:
    - taiga.org.com
### Parameters for the taiga service
###
taiga:

  ### The name on which the taiga backend API server will be accessible. In this chart, it's the same name you will use to
  ### access the frontend.  Notes:
  ### - Must be formatted exactly as you would type the hostname part of the URL in a browser, including the port number if needed.
  ### - Must be reachable by ALL clients using the service. See F.A.Q. in the readme, for more informations.
  ###
  apiserver: taiga.org.com

  ### Whether taiga is behind a TLS termination; if using a TLS ingress, set this to true.
  ###
  behindTlsProxy: true

  ### Database access parameters
  ### If deploying postgres, the dbHost should be `${release_name}-postgresql`
  ###
  dbHost: my-taiga-postgresql-0
  dbName: taiga
  dbUser: taiga
  dbPassword: "password"

  ### If you enable email, also uncomment and configure the parameters below
  ###
  emailEnabled: false
  # emailFrom: "taiga@mycompany.net"
  # emailUseTls: true
  # emailSmtpHost: "smtp.example.com"
  # emailSmtpPort: 587
  # emailSmtpUser: "user@example.com"
  # emailSmtpPassword: ""
  ldap:
    enabled: False
    use_tls: False
    ldap_server: ldap://<ldap_server_host_name>
    port: 389
    bind_dn: "cn=something,ou=admins,ou=theouhere,ou=auth,o=ethz,c=ch"
    base_dn: "basednhere"
    password: "changeme"
    username_attribute: "uid"
    email_attribute: "mail"
    full_name_attribute: "displayName"
    fallback: "normal"
    save_password: False
    filter: "(mail=*)"
  ### Configure taiga-back's secret key
  ### If omitted, a random alphanumeric string of 10 characters will be generated
  ###
  # secretKey: "changeme"


### Persistent volume and database configuration for the taiga service
###
persistence:

  ### Whether or not you want this chart to provision the database for you
  ### Database parameters must be specified under the `postgres` value
  ###
  deployPostgres: true

  ### If you don't want persistence for the taiga media storage, disable this
  ###
  enabled: true
  size: 8Gi
  accessMode: ReadWriteOnce
  annotations:
    "helm.sh/resource-policy": keep

  ## StorageClass for the provisioned PVC.
  ##
  storageClass: ""

  ### If you prefer using a PVC defined outside of this chart instead, uncomment this
  ### Requires `persistence.enabled: true`
  ###
  # existingClaim: ""


### Any stable/postgres values can be defined here
###
postgresql:

  ### Create a postgres user, password and database
  ###
  postgresqlUsername: "taiga"
  postgresqlPassword: "password"
  postgresqlDatabase: "taiga"

  persistence:
    enabled: true
    size: 2Gi
    annotations:
      "helm.sh/resource-policy": keep

    ### StorageClass for the provisioned PVC.
    ###
    storageClass: ""

    ### If you prefer using a PVC defined outside of this chart instead, uncomment this
    ### Requires `postgresql.persistence.enabled: true`
    ###
    # existingClaim: ""
```

**Comment by TheFuzzyTech** [comment]
Digging into this a little more, it looks like the container might not be able to resolve the hostname for the DB. I connected to the init-db container and tried `ping $TAIGA_DB_HOST` which is what checkdb.py pulls from env and it failed to resolve. I did verify that the env value was the name of the pod and the rest of the db settings were present in the env.

```root@nfv-taiga-7ff7dd5c6f-4tg4w:/usr/src/taiga-back# ping $TAIGA_DB_HOST
ping: nfv-taiga-postgresql-0: Temporary failure in name resolution```

**Comment by mvitale1989** [comment]
Hello, i can actually confirm that this is reproducible at least under certain conditions (tested on my k3s 1.20.4 installation), though i haven't dug yet into the why.

However i see that the postgres chart also install a headless service, which would definitely be a more proper way to access the db than what i documented in the chart. Can you try setting the dbHost varable to it? In your case it would be `dbHost: nfv-taiga-postgresql-headless` i believe.

**Comment by TheFuzzyTech** [comment]
Switching to the headless service doesn't seem to make a difference. The service doesn't have a cluster IP tied to it, though.

Container Logs
```kubectl logs nfv-taiga-86d8b74f9c-qdssx -c init-db
Templating out taiga and nginx configuration files
Enabling external SSL support! SSL handling must be done by a reverse proxy or a similar system
Waiting for DB to come up (timeout 60 seconds)...
```

Services
```
...
nfv-taiga-postgresql-headless   ClusterIP   None           <none>        5432/TCP   3m15s
nfv-taiga-postgresql            ClusterIP   10.43.133.90   <none>        5432/TCP   3m15s
nfv-taiga                       ClusterIP   10.43.47.128   <none>        80/TCP     3m15s
...
```

**Comment by mvitale1989** [comment]
Is your postgres container actually running and reachable? If it is, you can try using that Pod's IP as `dbHost`, as a test to bypass name resoution. You can get postgres' pod IP with e.g. this command `kubectl get pod --all-namespaces -o wide | grep postgresql
`. Also you can check that its IP appears within the headless service endpoint (e.g. `kubectl -n $TAIGA_NAMESPACE get endpoints nfv-taiga-postgresql-headless`)

If using the IP doesn't work either there may be a deeper issue with your cluster, e.g. missing CNI plugin installation.
If you instead see postgres as pending you're probably missing a storage provisioner within your cluster, which unless you provision PVs manually is required to enable persistence.

**Comment by RajPathare** [comment]
Hey @mvitale1989! I tried deploying this on my k8s cluster but I'm facing the same issue. There is proper connectivity in the cluster and I don't see any DNS resolution issues. The backend container never spins up. It gets stuck here - 

Waiting for DB to come up (timeout 0 seconds)...
Max sleep time reached while waiting for DB. Giving up.
Database is not yet reachable. Continuing execution anyway.
Starting taiga backend...
Registering SIGTERM handler...

As suggested, I tried changing "dbHost" to headless postgresql service but that doesn't work either.

**Comment by TheFuzzyTech** [comment]
I did some more digging over the weekend and DNS resolution looks to be working ok now, however, the init-db container still continues to fail like @RajPathare is stating.

**Comment by mvitale1989** [comment]
Mmh unfortunately I can't reproduce your issue. Setting `taiga.dbHost` to `${RELEASE_NAME}-postgresql-headless` fixes this issue in my test install, and init-db runs succesfully. Which k8s version are you running?

I would recommend the following steps as sanity checks:
1. Is the postgresql container running? e.g. `kubectl -n $RELEASE_NAMESPACE get pod -o wide "${RELEASE_NAME}-postgresql-0"`
2. Can you connect to the postgresql instance from inside of its container, through the TCP port? E.g. you can run `kubectl -n $RELEASE_NAMESPACE exec -it "${RELEASE_NAME}-postgresql-0" -- psql -c "\conninfo" "user=$DB_USERNAME password=$DB_PASSWORD host=127.0.0.1 port=5432"`
3. Can you connect to the postgresql instance from a different Pod? If you're running multiple k8s nodes, does this also work if the client and the postgresql Pods are on different nodes?
4. Is the IP of the postgres container in the headless service's endpoints list? Step 1 gave you the Pod IP, so you can check if that appears in the output of `kubectl -n $RELEASE_NAMESPACE` get endpoints "${RELEASE_NAME}-postgresql-headless"`
5. Does your init-db container actually point to the headless service, as you expect it to? Double check that running `echo $TAIGA_HOST_DB` in the init-db container outputs exactly `${RELEASE_NAME}-postgresql-headless`
6. Does pinging the headless service from the init-db container actually try to ping the postgres Pod's IP? E.g. try running `ping $TAIGA_DB_HOST`, and check that it matches the IP returned from step 1. You may want to increase the value of the `TAIGA_SLEEP` variable in the taiga deployment, to have more time to run commands there
6. Do you have any NetworkPolicies that prevent traffic among pods?

It would help to know which of these steps returns unexpected results to pinpoint the root cause

**Comment by puckettgw** [comment]
I have a similar issue -- however, my init-db container seems to find the database fine, but then taiga-back dies after failing to connect to the DB.

Also it appears that RELEASE_NAME env var is not set in the container, so I set dbHost to the expected service name within the values.yaml file - this is what allows init-db to work at all but taiga-back still fails.

**Comment by puckettgw** [comment]
Last few lines of init-db:
```
processing file django.po in /usr/src/taiga-back/taiga/locale/fi/LC_MESSAGES
processing file django.po in /usr/src/taiga-back/taiga/locale/ko/LC_MESSAGES
processing file django.po in /usr/src/taiga-back/taiga/locale/fr/LC_MESSAGES
Database checks completed.
Requested database-check only run. Exiting.
```

Logs from taiga-back:
```
Templating out taiga and nginx configuration files
Enabling external SSL support! SSL handling must be done by a reverse proxy or a similar system
Waiting for DB to come up (timeout 0 seconds)...
Max sleep time reached while waiting for DB. Giving up.
Database is not yet reachable. Continuing execution anyway.
Starting taiga backend...
Registering SIGTERM handler...
All background services have been started. Waiting on PIDs:  18
```
Eventually taiga-back container dies. It never is able to connect even when init-db can.

**Comment by puckettgw** [comment]
So now the init-db container is repeating this error:

```
postgresql 15:57:47.70 INFO  ==> ** Starting PostgreSQL **
2021-03-24 15:57:48.011 GMT [1] LOG:  pgaudit extension initialized
2021-03-24 15:57:48.012 GMT [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2021-03-24 15:57:48.012 GMT [1] LOG:  listening on IPv6 address "::", port 5432
2021-03-24 15:57:48.022 GMT [1] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
2021-03-24 15:57:48.059 GMT [133] LOG:  database system was shut down at 2021-03-24 15:57:47 GMT
2021-03-24 15:57:48.070 GMT [1] LOG:  database system is ready to accept connections
2021-03-24 15:57:48.593 GMT [140] FATAL:  password authentication failed for user "taiga"
2021-03-24 15:57:48.593 GMT [140] DETAIL:  Role "taiga" does not exist.
        Connection matched pg_hba.conf line 1: "host     all             all             0.0.0.0/0               md5"
2021-03-24 15:57:49.996 GMT [141] FATAL:  password authentication failed for user "taiga"
2021-03-24 15:57:49.996 GMT [141] DETAIL:  Role "taiga" does not exist.
```

**Comment by mvitale1989** [comment]
As this has been a source of confusion for many: i just released version 0.2.3 of the chart, in which don't have to specify `taiga.dbHost` by yourself anymore if you're using the embedded postgresql chart (which is most probably  the case for all of you).

@TheFuzzyTech  @RajPathare I suggest trying to remove what you specified in `taiga.dbHost` from your `values.yaml`, and try using this new chart version. Does the init-db container reach the database now?

**Comment by mvitale1989** [comment]
@puckettgw Those logs seem ok to me. Perhaps the line `Database is not yet reachable. Continuing execution anyway.` may be confusing: that's actually a lie, the database may just as well be there, because the taiga-back container didn't actually perform any database check before printing that line (see [this env var](https://github.com/mvitale1989/helm-taiga/blob/master/templates/deployment.yaml#L71-L72), and [its docs](https://github.com/mvitale1989/docker-taiga#extra-configuration-options)). Also, init-db and taiga-back run in the same Pod and share the same env vars with the db coordinates, meaning that if one of them can, the other almost certainly can as well.

So I believe the reason your taiga-back dies lies elsewhere. Do you have more logs, or more context? E.g. how long does the container stay alive? What does `kubectl describe` say the exit code, or reason for exiting, is?

> Also it appears that RELEASE_NAME env var is not set in the container, so I set dbHost to the expected service name within the values.yaml file - this is what allows init-db to work at all but taiga-back still fails.

Sorry for the confusion, in that case RELEASE_NAME was not meant to be an actual environment variable to use in the container. I meant that if you named your release `mytaiga`, your `taiga.dbHost` should be set to `mytaiga-postgresql-headless`. But as stated above the version i just released makes the suggestions i gave in that message obsolete.

**Comment by skye0402** [comment]
Same problem here with the 0.2.3 chart version. 
`
2021-04-02T11:07:32.823783453Z | Failed to connect to database server or database does not exist.
2021-04-02T11:07:32.746710491Z | Running database check
2021-04-02T11:07:32.74670439Z | Database is not yet reachable. Continuing execution anyway.
2021-04-02T11:07:32.746651689Z | Max sleep time reached while waiting for DB. Giving up.
2021-04-02T11:06:27.52019563Z | Waiting for DB to come up (timeout 60 seconds)...
2021-04-02T11:06:27.42333316Z | Templating out taiga and nginx configuration files
`
The DB is up and running in the first pod.

**Comment by mvitale1989** [comment]
@skye0402 can you show me your sanitized `values.yaml`?

Please also try installing the chart without specifying any values, does it work then? E.g. `helm -n taiga install mytaiga ./helm-taiga/` works out of the box for me, i cannot reproduce the fault condition.

**Comment by skye0402** [comment]
@mvitale1989 I just used the standard values.yaml as was provided from the cloning. Didn't edit it. Maybe I should have?

**Comment by mvitale1989** [comment]
@skye0402 if you're using 0.2.3, it's perfectly fine to deploy this without specifying any values.yaml. The default configuration should have taiga use the embedded postgres' headless service.

Again, I'm not able to reproduce this unfortunately. I would need to know exactly which steps of [this sanity check list](https://github.com/mvitale1989/helm-taiga/issues/7#issuecomment-790200944) fail for you, to gather some clues on what may be wrong. (NB: no need to specify any `taiga.dbHost` if you're using 0.2.3; also, the `${RELEASE_NAME}` you see there is not an actual environment variable, just write your release name in its place in every given command)

**Comment by skye0402** [comment]
Thank you for helping! I went through the sanity check list
```
kubectl get pod -o wide "apjio-taiga-postgresql-0"
NAME                       READY   STATUS    RESTARTS   AGE     IP             NODE                                                 NOMINATED NODE   READINESS GATES
apjio-taiga-postgresql-0   2/2     Running   0          4m31s   100.96.1.105   shoot--kyma--xxxx-cpu-worker-0-z3-b86d6-4gnch   <none>           <none>
```
and 
```
kubectl exec -it "apjio-taiga-postgresql-0" -- psql -c "\conninfo" "user=taiga password=changeme host=127.0.0.1 port=5432"                         Defaulting container name to apjio-taiga-postgresql.
Use 'kubectl describe pod/apjio-taiga-postgresql-0 -n taiga-helm' to see all of the containers in this pod.
You are connected to database "taiga" as user "taiga" on host "127.0.0.1" at port "5432".
```
and
```
kubectl get endpoints apjio-taiga-postgresql-headless
NAME                              ENDPOINTS           AGE
apjio-taiga-postgresql-headless   100.96.1.105:5432   7m27s
```
So I think the db instance is ok? However, the other pod doesn't come up as it can't find the DB. As it is in an init state I don't know how I can get more details.

```
NAME                          READY   STATUS       RESTARTS   AGE
apjio-taiga-58d69bc7c-v922p   0/3     Init:Error   2          4m14s
```

**Comment by mvitale1989** [comment]
Yes, the db container looks good to me as well.

So probably the taiga container itself is pointing to the wrong database endpoint? Theoretically it should point to the postgresql headless service for this to work, so `apjio-taiga-postgresql-headless` in your case.
You can check the value of `TAIGA_DB_HOST` in the configmap containing the various env vars. In your case i believe you can check its content with `kubectl describe configmap apjio-taiga`. What is the variable value in your case?

**Comment by skye0402** [comment]
I will check that thank you indeed! I got my setup (based on the official docker-compose.yml) running meanwhile - however (as I'm not very knowledgeable) the TAIGA_URL (which is localhost in their standard docker-compose.yml) is called by the browser - which means it can't be localhost but must be the external URL of the pods in which the front and back are running. I can set this at runtime but this is a very bad idea I think. I wonder why it doesn't get abstracted by the nginx gateway.
How did you solve this in your setup? I'm very happy to share mine once it works by the way (or beforehand if it helps).

**Comment by mvitale1989** [comment]
I'm not familiar with the docker-compose file you're talking about, but i assume that `TAIGA_URL` is what i called `TAIGA_HOSTNAME` in [my taiga docker image](https://github.com/mvitale1989/docker-taiga#extra-configuration-options).

The reason you need to explicitly tell taiga where it's located (and cannot simply rely on nginx itself telling taiga where it is) is that sometimes, taiga needs to perform operations non-interactively. Sure, when you open taiga in the browser, the backend will just look at the `Host` header and figure out its external URL. But what if needs to generate an email containing a link to a task? In that case taiga only has its configuration to look at, so to produce a valid URL taiga needs to know how it can be reached externally. So it takes that information from the `TAIGA_URL` env var.

**Comment by mvitale1989** [comment]
Closing this, as the original issue should be solved with 0.2.3
